### PR TITLE
Refactor OmniClient balance types

### PR DESCRIPTION
**Pull request by msgilligan**
Refactor OmniClient balance types

* Rename MPBalanceEntry to BalanceEntry
* BalanceEntry replaces ConsensusEntry
* Subclasses of BalanceEntry for specific RPCs (include Address or Property)
* CurrencyID is comparable (for use as Map index)
* Custom deserializers to generate sorted Maps of BalanceEntry
* Many new test Specs

### CODE REVIEW

**Comment by msgilligan** [comment]
@dexX7 Sorry it took me three tries to get this pushed correctly.

File: omnij-rpc/src/integ/groovy/foundation/omni/test/rpc/basic/GetBalancesSpec.groovy [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import spock.lang.Stepwise
+
+import static foundation.omni.CurrencyID.MSC
+import static foundation.omni.CurrencyID.TMSC
+
+
+/**
+ * Fund an address and check its balances in three different ways
+ */
+class GetBalancesSpec extends BaseRegTestSpec {
```
Comment (msgilligan) [inline omnij-rpc/src/integ/groovy/foundation/omni/test/rpc/basic/GetBalancesSpec.groovy]: @dexX7 This one tests all three balance checking APIs and demonstrates how they work.

File: omnij-rpc/src/main/groovy/foundation/omni/consensus/OmniCoreConsensusTool.groovy [hunk 2]
```diff
@@ -58,25 +58,10 @@ class OmniCoreConsensusTool extends ConsensusTool {
...
-                map.put(address, entry)
-            }
-        }
-        return map;
-    }
-
-    private ConsensusEntry itemToEntry(MPBalanceEntry item) {
-        return new ConsensusEntry(balance: item.balance, reserved:item.reserved)
+    private SortedMap<Address, BalanceEntry> getConsensusForCurrency(CurrencyID currencyID) {
+        SortedMap<Address, BalanceEntry> balances = client.omniGetAllBalancesForId(currencyID)
```
Comment (msgilligan) [inline omnij-rpc/src/main/groovy/foundation/omni/consensus/OmniCoreConsensusTool.groovy]: @dexX7 Note that we eliminated a data copy/convert of all the balances here.

File: omnij-rpc/src/main/java/foundation/omni/json/conversion/AddressBalanceEntriesDeserializer.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import foundation.omni.rpc.AddressBalanceEntries;
+import foundation.omni.rpc.AddressBalanceEntry;
+import foundation.omni.rpc.BalanceEntry;
+
+import java.io.IOException;
+
+/**
+ *
+ */
+public class AddressBalanceEntriesDeserializer extends StdDeserializer<AddressBalanceEntries> {
```
Comment (msgilligan) [inline omnij-rpc/src/main/java/foundation/omni/json/conversion/AddressBalanceEntriesDeserializer.java]: Figuring out how to do this via limited Jackson documentation took me more time than I'd like to admit.

**Comment by msgilligan** [comment]
@dexX7 If you can easily run consensus tests locally with this PR that would be helpful. :)

File: omnij-rpc/src/integ/groovy/foundation/omni/test/rpc/basic/GetBalancesSpec.groovy (hunk 1 above)
Comment (dexX7) [inline omnij-rpc/src/integ/groovy/foundation/omni/test/rpc/basic/GetBalancesSpec.groovy]: Nice!

**Comment by dexX7** [comment]
There are 50 failures, which are not caused my block count mismatch:

http://bitwatch.co/uploads/omnij-67415952f4/

File: omnij-rpc/src/main/groovy/foundation/omni/consensus/OmniCoreConsensusTool.groovy [hunk 4]
```diff
@@ -58,25 +58,10 @@ class OmniCoreConsensusTool extends ConsensusTool {
...
-        List<MPBalanceEntry> balances = client.omniGetAllBalancesForId(currencyID)
-
-        TreeMap<Address, ConsensusEntry> map = [:]
-
-        balances.each { MPBalanceEntry item ->
-
-            Address address = item.address
-            ConsensusEntry entry = itemToEntry(item)
-
-            if (address != "" && entry.balance > 0) {
```
Comment (msgilligan) [inline omnij-rpc/src/main/groovy/foundation/omni/consensus/OmniCoreConsensusTool.groovy]: @dexX7 I wonder if removing the `address != "" && entry.balance > 0` filter here is what caused the 50 failures?

Comment (msgilligan) [inline omnij-rpc/src/main/groovy/foundation/omni/consensus/OmniCoreConsensusTool.groovy]: Similar filters remain for `OmniwalletConsensusTool` and `ChestConsensusTool`. Should we remove those? Could you try that locally?

Comment (dexX7) [inline omnij-rpc/src/main/groovy/foundation/omni/consensus/OmniCoreConsensusTool.groovy]: Ha! Omni Core doesn't return emtpy balances, but OmniChest and OmniWallet do.

The issue is now clearer, especially given the test report. The extra filtering here is needed, but it should be:
```diff
diff --git a/omnij-rpc/src/main/groovy/foundation/omni/consensus/ChestConsensusTool.groovy b/omnij-rpc/src/main/groovy/foundation/omni/consensus/ChestConsensusTool.groovy
index 4e10fb2..5fd23fe 100644
--- a/omnij-rpc/src/main/groovy/foundation/omni/consensus/ChestConsensusTool.groovy
+++ b/omnij-rpc/src/main/groovy/foundation/omni/consensus/ChestConsensusTool.groovy
@@ -51,3 +51,3 @@ class ChestConsensusTool extends ConsensusTool {
 
-            if (address != "" && entry.balance > 0) {
+            if (address != "" && (entry.balance > 0 || entry.reserved > 0)) {
                 map.put(address, entry)
diff --git a/omnij-rpc/src/main/groovy/foundation/omni/consensus/OmniwalletConsensusTool.groovy b/omnij-rpc/src/main/groovy/foundation/omni/consensus/OmniwalletConsensusTool.groovy
index 96bc46e..c9f656e 100644
--- a/omnij-rpc/src/main/groovy/foundation/omni/consensus/OmniwalletConsensusTool.groovy
+++ b/omnij-rpc/src/main/groovy/foundation/omni/consensus/OmniwalletConsensusTool.groovy
@@ -48,3 +48,3 @@ class OmniwalletConsensusTool extends ConsensusTool {
 
-            if (address != "" && entry.balance > 0) {
+            if (address != "" && (entry.balance > 0 || entry.reserved > 0)) {
                 map.put(address, entry)
```

Up until now, we basically trimmed balance entries, which only have a reserved balance.

**Comment by dexX7** [comment]
Tests passed! :)

Except:
```
groovy.json.JsonException: Unable to process url: https://blockchain.info/latestblock
// was rejected
```

But this is unrelated. Might be useful to use a reliable API, which isn't blockchain.info, though; for example [blockcypher.com](http://dev.blockcypher.com/#chain-endpoint).

**Comment by msgilligan** [comment]
I added [BlockCypherSyncing.groovy](https://github.com/msgilligan/bitcoinj-addons/blob/master/bitcoinj-rpcclient/src/main/groovy/com/msgilligan/bitcoinj/rpc/BlockCypherSyncing.groovy) to `bitcoinj-addons`. Once that version of `bitcoinj-addons` is released we can change `BlockchainDotInfoSyncing` to `BlockCypherSyncing` in [BaseMainNetSpec.groovy](https://github.com/OmniLayer/OmniJ/blob/master/omnij-rpc/src/integ/groovy/foundation/omni/BaseMainNetSpec.groovy#L23). I don't think their rate-limiting would be a problem -- if it is we'd need to add the API token somehow.

**Comment by msgilligan** [comment]
So, by removing the filter from `OmniCoreConsensusTool` and changing the filter in `OmniwalletConsensusTool` and `ChestConsensusTool` we've increased our test coverage to include addresses with zero available balance but a reserved balance.

Merging...

**Comment by dexX7** [comment]
blockchain.info has a rate limit, too, although it's much higher. I hope we won't hit it, because I think it would be better to get rid of blockchain.info due to it's non-availability from time to time, and worse, it's messed up code, which doesn't necessarily follow the chain, which Bitcoin Core considers as valid (e.g. during the BIP66 switch over). I'm a bit dramatic here, but still..

+1 for the merge.

Once 0.0.10 is out, we could tackle the consensus hashing. Right now in Omni Core we use a hash-over-balances for checkpoints, and @zathras-crypto created https://github.com/OmniLayer/omnicore/pull/226 some time ago to cover the whole state (although I believe it still needs some tweaking, potentially hand in hand moving parts in the block handlers, because it's currently not very doable to traverse the chain backwards and check the consensus hashes on the way - which would be extremely nice! If we are at this point, then we could simplify the reorg tests greatly, by creating some transactions, saving the hashes, and then reverting all blocks).

**Comment by msgilligan** [comment]
> I think it would be better to get rid of blockchain.info due to it's non-availability from time to time

Another option would be to use a **bitcoinj** `PeerGroup` to talk directly to the P2P network to get the block height.

**Comment by dexX7** [comment]
In general this is a very interesting idea, but probably an overkill for this purpose?

**Comment by msgilligan** [comment]
Well when I originally thought of it, it was definitely overkill. But all it would take is a Spring container and [PeerGroupService.java](https://github.com/msgilligan/bitcoinj-addons/blob/master/bitcoinj-server/src/main/java/com/msgilligan/bitcoinj/spring/service/PeerGroupService.java) or it could be done directly with the **bitcoinj** `PeerGroup`. 

Since we'll eventually be using/building unit test fixtures for Spring components like `PeerGroupService`, it will get even easier.

I'm not suggesting we do at this time, but it was worth mentioning.

**Comment by msgilligan** [comment]
@dexX7 

> Ha! Omni Core doesn't return emtpy balances, but OmniChest and OmniWallet do.

That's not what I was seeing in [Jenkins Job 220](https://ci.omni.foundation/job/omni-mainnet-consensus/220/). I fixed it with 2e5fbb4aba4c00330450155a37a2537244c5d307.

Was that the right thing to do?

**Comment by dexX7** [comment]
Ahh @msgilligan, good catch!

You're right. When using `"omni_getallbalancesforid"`, then there may be entries with zero balances. Those are entries, which had some balance, but now have none. `"omni_getallbalancesforaddress"` has a similar behavior.

I don't think this is desired, so it should probably be fixed.

2e5fbb4aba4c00330450155a37a2537244c5d307 looks good to me.

**Comment by msgilligan** [comment]
> 2e5fbb4 looks good to me.

Although we should (eventually) remove that and write a unit test that catches the error (https://github.com/OmniLayer/omnicore/issues/257). Right?

**Comment by dexX7** [comment]
Agreed. Once 257 is handled, we could do so for (local) Omni Core. And after the release, and once hopefully all remote servers have upgraded, the additional checks could be removed completely.

**Comment by dexX7** [comment]
@msgilligan: https://github.com/OmniLayer/omnicore/pull/262 was merged (though not included in the binary OmniJ uses, but this is probably a different story).

**Pull request merged**